src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -444,16 +446,28 @@ joinOuter :: MonadIO m =>
 joinOuter eq s1 s =
     Stream.concatM $ do
         arr <- Array.fromStream $ fmap (,False) s
-        return $ go arr <> leftOver arr
+        let len = length arr
+        arr2 <-
+            Stream.fold
+            (MA.writeN len)
+            (Stream.fromList (Prelude.replicate len False))
+        return $ go arr arr2  <> leftOver arr arr2
```

Comment:
This looks unnecessarily complicated.

Why not just use the existing code with following modifications:

* Make the existing Array a mutable array
* Uncomment line 476 in the old code, and use `modifyIndex` to modify the flag in the tuple to True.